test/jdk/tools/jar/JarExtractTest.java

```diff
@@ -0,0 +1,382 @@
+        Assert.assertEquals(Files.readAllBytes(f2), FILE_CONTENT, "Unexpected content in file " + f2);
+    }
+
+    /**
+     * Tests that the {@code -P} option cannot be used during jar extraction when the {@code -C} and/or
+     * {@code --dir} option is used
+     */
+    @Test
+    public void testExtractWithDirPFlagNotAllowed() throws Exception {
+        final String expectedErrMsg = "-P option cannot be used when extracting a jar to a specific location";
+        final String tmpDir = Files.createTempDirectory(Path.of("."), "8173970-").toString();
+        final List<String[]> cmdArgs = new ArrayList<>();
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "-C", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "--dir", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "-C", "."});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "--dir", "."});
+        cmdArgs.add(new String[]{"-xvfP", testJarPath.toString(), "-C", tmpDir});
+        for (final String[] args : cmdArgs) {
+            final ByteArrayOutputStream err = new ByteArrayOutputStream();
+            printJarCommand(args);
+            int exitCode = JAR_TOOL.run(System.out, new PrintStream(err), args);
+            Assert.assertNotEquals(exitCode, 0, "jar extraction was expected to fail but didn't");
+            // verify it did indeed fail due to the right reason
+            Assert.assertTrue(err.toString(StandardCharsets.UTF_8).contains(expectedErrMsg));
+        }
+    }
+
+    /**
+     * Tests that {@code jar -xvf <jarname> -C <dir>} works fine too
+     */
+    @Test
+    public void testLegacyCompatibilityMode() throws Exception {
+        final String tmpDir = Files.createTempDirectory(Path.of("."), "8173970-").toString();
+        final String[] args = new String[]{"-xvf", testJarPath.toString(), "-C", tmpDir};
+        printJarCommand(args);
+        final int exitCode = JAR_TOOL.run(System.out, System.err, args);
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + testJarPath);
+        verifyExtractedContent(tmpDir);
+    }
+
+    /**
+     * Tests that when multiple directories are specified for extracting the jar, the jar extraction
+     * fails
+     */
+    @Test
+    public void testExtractFailWithMultipleDir() throws Exception {
+        final String expectedErrMsg = "You may not specify more than one directory for extracting the jar";
+        final String tmpDir = Files.createTempDirectory(Path.of("."), "8173970-").toString();
+        final List<String[]> cmdArgs = new ArrayList<>();
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-C", tmpDir, "-C", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "--dir", tmpDir, "--dir", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "--dir", tmpDir, "-C", tmpDir});
```

Comment:
Perhaps use a DataProvider so you can test --extract as well?

Reply:
Done